crates/cargo-contract/src/cmd/extrinsics/balance.rs

```diff
@@ -89,16 +91,31 @@ impl TokenMetadata {
             .as_str()
             .context("error converting symbol to string")?;
         Ok(Self {
-            denomination,
+            token_decimals,
             symbol: symbol.to_string(),
         })
     }
 }
 
-impl TryFrom<String> for DenominatedBalance {
-    type Error = anyhow::Error;
+impl FromStr for BalanceVariant {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let input = input.replace('_', "");
+        if input.contains('.') || input.ends_with(|ch: char| ch.is_alphabetic()) {
```

Comment:
Could you please elaborate where the ambiguity is caused here? I assumed if a user explicitly specifies a value with a decimal point, this already implies that denominated units are used. It would be odd if users normally enter `1_000_000.0`. Nonetheless, if user decided to add `.0` at the end, it will still work fine, the string will go through the parsing and result in `1_u128` when encoding

Reply:
The conditional `if input.contains('.') || input.ends_with(|ch: char| ch.is_alphabetic())` gives the impression that just a decimal point is enough to specify a denominated value.

However I just ran a test:
```
assert_eq!(
    BalanceVariant::from_str("1").unwrap(),
    BalanceVariant::from_str("1.0").unwrap()
)
```
And the second one fails because `DenominatedBalance::from_str` in fact does require a token symbol. So it is okay then, requires a test case perhaps.

The conditional here is a bit confusing because it is doing sort of a precheck that it *might* be Denominated balance and then trying to parse it. Alternative might be to do something like:

```
        Ok(match input.parse::<Balance>() {
            Ok(balance) => BalanceVariant::Default(balance),
            Err(_) => BalanceVariant::Denominated(DenominatedBalance::from_str(
                &input,
            )?)
        })
```

Reply:
Okay. That makes sense, I will rewrite and separate parsing logic